### PR TITLE
[Project] Add method to stabilize UUIDs

### LINKED ISSUE #175: Deterministic UUIDs
I have been thinking about doing something similar to this for a while and the [xUnique](https://github.com/truebit/xUnique) revived my interest on this feature.

##### Context

The algorithm used by Xcode to generate UUIDs is not known. Xcodeproj currently has a conservative approach at tries to make UUIDs which are similar to those generated by Xcodeproj while checking if a duplicate already exists in the project before using a new one. A duplicate UUID would corrupt the project.

[xUnique](https://github.com/truebit/xUnique) proves that Xcode accepts UUIDs regardless of the length and of the type.

##### Proposal

- Generate longer UUIDs
- Derive the UUID from a hash of the attributes of the new objects and the UUID of the parents. This would greatly reduce the noise when recreating a project from scratch.
- Add a subcommand to “normalize” projects.

##### Alternative

- Prefix UUIDs with a string to namespace the ones generated by Xcodeproj and avoid the need to check for collisions.
- Use a hash of the timestamp for the UUIDs.

### PR DESCRIPTION
**Pull request by segiddins**
Closes #175

### CODE REVIEW

**Comment by segiddins** [comment]
\c @alloy @kylef @orta @neonichu @mrackwitz 

I know the diff is small, and it needs some more tests, but I think it works.

**Comment by orta** [comment]
cool!

**Comment by alloy** [comment]
Awesome!

Yeah I'd like to see some tests for various types of objects.

The doc should also state that it changes the uuids of objects and thus the user should not depend on being able to find them by uuid afterwards.

**Comment by neonichu** [comment]
Awesome :tada:

**Comment by zats** [comment]
Hi, I'm sorry to intrude, just wanted to mention that original Xcode `PBXObject` ID generation can be found under DevToolsSupport.framework `+[NSString stringWithHexadecimalRepresentationOfUniqueIdentifier]` if that is of any relevance :flushed: 

And a question regarding your logic of UUID generation: if I have two references pointing to the same file in the project, does it mean the generated UUID will be the same for both of them? 
Thanks!

**Comment by segiddins** [comment]
@zats the issue with Xcode's UUID generation is that it's _meant_ to be unique -- it includes a time stamp component.

As to your question, they shouldn't be -- thanks to the distinction between PBXFileReferences and PBXBuildFiles, this should be fine.

**Comment by zats** [comment]
oh I see, thanks for the explanation!

File: lib/xcodeproj/project.rb [hunk 1]
```diff
@@ -409,6 +410,12 @@ def uuids
       objects_by_uuid.keys
     end
 
+    # @note this should only be used for entirely machine-generated projects
+    #
+    def predictabilize_uuids
```
Comment (kylef) [inline lib/xcodeproj/project.rb]: Is there a counter-part PR using this to CocoaPods?

Comment (segiddins) [inline lib/xcodeproj/project.rb]: yes

Comment (segiddins) [inline lib/xcodeproj/project.rb]: https://github.com/CocoaPods/CocoaPods/pull/3750

**Comment by mrousavy** [comment]
I think I'm experiencing an issue with deterministic UUIDs [here](https://github.com/CocoaPods/CocoaPods/issues/10506), do you have an idea what might be wrong here @segiddins ?

**Pull request merged**